From 58ae9bbbba1c0fc7e4673bbda1805875890b4772 Mon Sep 17 00:00:00 2001
From: Lars-Peter Clausen <lars@metafoo.de>
Date: Fri, 3 Jul 2015 17:03:33 +0200
Subject: [PATCH 13/16] [media] adv7604: Fix HDMI mode detection for the
 ADV7611

The ADV7611 uses a different interrupt bit and register than the ADV7611.
Add support for this.

Signed-off-by: Lars-Peter Clausen <lars@metafoo.de>
---
 drivers/media/i2c/adv7604.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/drivers/media/i2c/adv7604.c b/drivers/media/i2c/adv7604.c
index b271a243a2fa..f6737f9077a0 100644
--- a/drivers/media/i2c/adv7604.c
+++ b/drivers/media/i2c/adv7604.c
@@ -2229,6 +2229,7 @@ static int adv76xx_isr(struct v4l2_subdev *sd, u32 status, bool *handled)
 	const u8 irq_reg_0x70 = io_read(sd, 0x70);
 	u8 fmt_change_digital;
 	u8 fmt_change;
+	u8 hdmi_mode;
 	u8 tx_5v;
 
 	if (irq_reg_0x43)
@@ -2256,8 +2257,17 @@ static int adv76xx_isr(struct v4l2_subdev *sd, u32 status, bool *handled)
 		if (handled)
 			*handled = true;
 	}
+
+	if (info->type == ADV7604) {
+		hdmi_mode = irq_reg_0x6b & 0x1;
+	} else {
+		hdmi_mode = io_read(sd, 0x66);
+		io_write(sd, 0x67, hdmi_mode);
+		hdmi_mode &= 0x08;
+	}
+
 	/* HDMI/DVI mode */
-	if (irq_reg_0x6b & 0x01) {
+	if (hdmi_mode) {
 		v4l2_dbg(1, debug, sd, "%s: irq %s mode\n", __func__,
 			(io_read(sd, 0x6a) & 0x01) ? "HDMI" : "DVI");
 		set_rgb_quantization_range(sd);
@@ -2914,6 +2924,7 @@ static void adv7604_setup_irqs(struct v4l2_subdev *sd)
 static void adv7611_setup_irqs(struct v4l2_subdev *sd)
 {
 	io_write(sd, 0x41, 0xd0); /* STDI irq for any change, disable INT2 */
+	io_write(sd, 0x69, 0x08); /* HDMI mode interrupt */
 }
 
 static void adv7612_setup_irqs(struct v4l2_subdev *sd)
-- 
2.17.1

